### PR TITLE
add stable diffusion example

### PR DESCRIPTION
**Pull request by prathikr**
# What does this PR do?

This PR adds an example of finetuning a diffusion model for the text to image task.

### CODE REVIEW

**Comment by prathikr** [comment]
@JingyaHuang can you please review?

**Review by JingyaHuang — COMMENTED** [review]
Hi @prathikr, it's awesome to add stable diffusion example! Thanks!

I left some nits, and when testing the script I got the following error:
```
Traceback (most recent call last):
  File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/onnxruntime/training/ortmodule/_utils.py", line 316, in get_exception_as_string
    raise exception
  File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/onnxruntime/training/ortmodule/_graph_execution_manager.py", line 425, in _get_exported_model
    torch.onnx.export(
  File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 506, in export
    _export(
  File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 1548, in _export
    graph, params_dict, torch_out = _model_to_graph(
  File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 1117, in _model_to_graph
    graph = _optimize_graph(
  File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 665, in _optimize_graph
    graph = _C._jit_pass_onnx(graph, operator_export_type)
  File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 1901, in _run_symbolic_function
    raise errors.UnsupportedOperatorError(
torch.onnx.errors.UnsupportedOperatorError: Exporting the operator 'aten::scaled_dot_product_attention' to ONNX opset version 15 is not supported. Please feel free to request support or submit a pull request on PyTorch GitHub: https://github.com/pytorch/pytorch/issues.
``` 
It seems to be an opset issue while exporting the ONNX graph. Can you confirm the version of onnx to use and probably add it to the requirements? Thanks.

File: examples/onnxruntime/training/stable-diffusion/text-to-image/README.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  --resolution=512 --center_crop --random_flip \
+  --train_batch_size=1 \
+  --gradient_accumulation_steps=4 \
+  --gradient_checkpointing \
+  --max_train_steps=15000 \ 
+  --learning_rate=1e-05 \
+  --max_grad_norm=1 \
+  --lr_scheduler="constant" --lr_warmup_steps=0 \
+  --output_dir="sd-pokemon-model" 
+```
```
[the remark this answers is not in the document]
Comment (prathikr) [reply examples/onnxruntime/training/stable-diffusion/text-to-image/README.md:136]: Added stats at the beginning.

File: examples/onnxruntime/training/stable-diffusion/text-to-image/README.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+```bash
+huggingface-cli login
+```
+
+If you have already cloned the repo, then you won't need to go through these steps.
+
+<br>
+
+#### Hardware
+With `gradient_checkpointing` and `mixed_precision` it should be possible to fine tune the model on a single 24GB GPU. For higher `batch_size` and faster training it's better to use GPUs with >30GB memory.
```
[the remark this answers is not in the document]
Comment (prathikr) [reply examples/onnxruntime/training/stable-diffusion/text-to-image/README.md:52]: Updated.

**Comment by prathikr** [comment]
> Hi @prathikr, it's awesome to add stable diffusion example! Thanks!
> 
> I left some nits, and when testing the script I got the following error:
> 
> ```
> Traceback (most recent call last):
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/onnxruntime/training/ortmodule/_utils.py", line 316, in get_exception_as_string
>     raise exception
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/onnxruntime/training/ortmodule/_graph_execution_manager.py", line 425, in _get_exported_model
>     torch.onnx.export(
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 506, in export
>     _export(
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 1548, in _export
>     graph, params_dict, torch_out = _model_to_graph(
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 1117, in _model_to_graph
>     graph = _optimize_graph(
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 665, in _optimize_graph
>     graph = _C._jit_pass_onnx(graph, operator_export_type)
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 1901, in _run_symbolic_function
>     raise errors.UnsupportedOperatorError(
> torch.onnx.errors.UnsupportedOperatorError: Exporting the operator 'aten::scaled_dot_product_attention' to ONNX opset version 15 is not supported. Please feel free to request support or submit a pull request on PyTorch GitHub: https://github.com/pytorch/pytorch/issues.
> ```
> 
> It seems to be an opset issue while exporting the ONNX graph. Can you confirm the version of onnx to use and probably add it to the requirements? Thanks.

@JingyaHuang this example currently required PyTorch

> Hi @prathikr, it's awesome to add stable diffusion example! Thanks!
> 
> I left some nits, and when testing the script I got the following error:
> 
> ```
> Traceback (most recent call last):
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/onnxruntime/training/ortmodule/_utils.py", line 316, in get_exception_as_string
>     raise exception
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/onnxruntime/training/ortmodule/_graph_execution_manager.py", line 425, in _get_exported_model
>     torch.onnx.export(
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 506, in export
>     _export(
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 1548, in _export
>     graph, params_dict, torch_out = _model_to_graph(
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 1117, in _model_to_graph
>     graph = _optimize_graph(
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 665, in _optimize_graph
>     graph = _C._jit_pass_onnx(graph, operator_export_type)
>   File "/home/onnxruntimedev/miniconda3/lib/python3.9/site-packages/torch/onnx/utils.py", line 1901, in _run_symbolic_function
>     raise errors.UnsupportedOperatorError(
> torch.onnx.errors.UnsupportedOperatorError: Exporting the operator 'aten::scaled_dot_product_attention' to ONNX opset version 15 is not supported. Please feel free to request support or submit a pull request on PyTorch GitHub: https://github.com/pytorch/pytorch/issues.
> ```
> 
> It seems to be an opset issue while exporting the ONNX graph. Can you confirm the version of onnx to use and probably add it to the requirements? Thanks.

@JingyaHuang please retry with PyTorch and ONNX Runtime nightly packages. I added a note in the readme about requiring nightly packages.

File: examples/onnxruntime/training/stable-diffusion/text-to-image/README.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  --lr_scheduler="constant" --lr_warmup_steps=0 \
+  --output_dir="sd-pokemon-model"
+```
+
+
+Once the training is finished the model will be saved in the `output_dir` specified in the command. In this example it's `sd-pokemon-model`. To load the fine-tuned model for inference just pass that path to `ORTStableDiffusionPipeline`
+
+
+```python
+from optimum import ORTStableDiffusionPipeline
```
Comment (JingyaHuang) [inline examples/onnxruntime/training/stable-diffusion/text-to-image/README.md:101]:
```suggestion
from optimum.onnxruntime import ORTStableDiffusionPipeline
```

File: examples/onnxruntime/training/stable-diffusion/text-to-image/README.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+## Running locally with PyTorch
+### Installing the dependencies
+
+___Note___: This example requires PyTorch nightly and [ONNX Runtime](https://github.com/Microsoft/onnxruntime) nightly
+```bash
+pip install --pre torch torchvision --index-url https://download.pytorch.org/whl/nightly/cu118
+pip install onnxruntime-training --pre -f https://download.onnxruntime.ai/onnxruntime_nightly_cu118.html
+python -m onnxruntime.training.ortmodule.torch_cpp_extensions.install
+```
```
Comment (JingyaHuang) [inline examples/onnxruntime/training/stable-diffusion/text-to-image/README.md:21]: Or get your environment ready via Docker: [`examples/onnxruntime/training/docker/Dockerfile-ort-nightly-cu118`](https://github.com/huggingface/optimum/blob/main/examples/onnxruntime/training/docker/Dockerfile-ort-nightly-cu118)

**Review by JingyaHuang — APPROVED** [review]
Hi @prathikr, thanks for the updates.

Tried with ort nightly, the script works fine (though massive log info looks a little scary, but guess it might be intended as it's still experimental). Left two nits before merging it.

Also,I think of adding it to our [ORT training doc](https://huggingface.co/docs/optimum/onnxruntime/usage_guides/trainer), we can add it in a separate PR (are you interested in it?)

File: examples/onnxruntime/training/stable-diffusion/text-to-image/README.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
```
Comment (JingyaHuang) [inline examples/onnxruntime/training/stable-diffusion/text-to-image/README.md:22]:
```suggestion
Or get your environment ready via Docker: [examples/onnxruntime/training/docker/Dockerfile-ort-nightly-cu118](https://github.com/huggingface/optimum/blob/main/examples/onnxruntime/training/docker/Dockerfile-ort-nightly-cu118)

```

**Comment by prathikr** [comment]
> Hi @prathikr, thanks for the updates.
> 
> Tried with ort nightly, the script works fine (though massive log info looks a little scary, but guess it might be intended as it's still experimental). Left two nits before merging it.
> 
> Also,I think of adding it to our [ORT training doc](https://huggingface.co/docs/optimum/onnxruntime/usage_guides/trainer), we can add it in a separate PR (are you interested in it?)

@JingyaHuang yes, let's follow up offline and I can work on a documentation PR.

**Pull request merged**